File: js_edition/src/api.spec.js
import assert from 'assert';
import { members, absences } from './api';

const everyItemContainsKey = (key) => (collection) =>
  collection.forEach((item) => assert(Object.keys(item).includes(key)));

describe('members', () => {
  describe('every member has key', () => {
    [
      'id',
      'name',
      'userId',
      'image',
    ].forEach((key) => {
      it(key, () => members().then(everyItemContainsKey(key)));
    });
  });
});

describe('absences', () => {
  describe('every absence has key', () => {
    [
      'admitterNote',
      'confirmedAt',
      'createdAt',
      'crewId',
      'endDate',
      'id',
      'memberNote',
      'rejectedAt',
      'startDate',
      'type',
      'userId',
    ].forEach((key) => {
      it(key, () => absences().then(everyItemContainsKey(key)));
    });
  });
});

File: ruby_edition/spec/api_spec.rb
require_relative '../cm_challenge/api'

RSpec.describe CmChallenge::Api do
  describe '#members' do
    let(:members) { CmChallenge::Api.members }

    it { expect(members).to respond_to(:each) }
    it { expect(members.length).to satisfy { |v| v >= 1 }  }
    it { expect(members).to all(have_key(:id)) }
    it { expect(members).to all(have_key(:name)) }
    it { expect(members).to all(have_key(:user_id)) }
  end

  describe '#absences' do
    let(:absences) { CmChallenge::Api.absences }

    it { expect(absences).to respond_to(:each) }
    it { expect(absences.length).to satisfy { |v| v >= 1 }  }
    it { expect(absences).to all(have_key(:id)) }
    it { expect(absences).to all(have_key(:start_date)) }
    it { expect(absences).to all(have_key(:end_date)) }
    it { expect(absences).to all(have_key(:user_id)) }
    it { expect(absences).to all(have_key(:crew_id)) }
  end
end
